### PR TITLE
Refs #23200 - Moving inline code to external scss files

### PR DESCRIPTION
**Pull request by ik5**
This stage contains the "bootstrap" for all the node stack, nothing more

### CODE REVIEW

File: package.json
Comment (ohadlevy) [inline package.json]: I would suggest not to load these unless you are adding code that uses it.

Comment (ik5) [reply package.json]: I wish to start moving stuff slowly to react, that's why I'ved added them, at the moment I want to start with supporting of scss loading using `.js` file.

**Comment by ik5** [comment]
Okay it looks great (from my side), but one thing is that I'm not sure here, does this package requires the tranformation and polyfills of babel, or will it be implemented by the main foreman asset manager.

**Comment by lzap** [comment]
Thanks, honestly this patch is terrifying for me. I don't understand much of JavaScript and I foresee packaging and production issues with that. Are you willing to do all the packaging changes and production testing?

But first let's get discovery tests fixed, then merge most of the other PRs and then let's talk more about this.

Also this change should pull all the javascript which we have, which is exactly one inline block. I am not really sold in this.

Why you use Refs instead Fixes? I see a separate issue for this. Should be Fixed imo.

**Comment by ik5** [comment]
@lzap the changes are compatible with the way foreman is going. As I understand only new things or dependencies that obsolete are moving from jQuery to React, while I was starting to add features I have already found such cases.

My task here is first to create such migration, and then the second task is to separate the inline Javascript and the inline CSS code into an external pipeline rather then Rails asset pipeline the way Foreman is currently working (aka stage 2).

Note that at this point, the every Javascript and CSS changes are inline rather then using external files, so it will create a much cleaner code that easier to maintain in the long run, and create a bit more DRY code in a whole.

The actual dependencies that I'm using, are also installed by foreman, and at this time I'm trying to figure out if I need to add that dependencies also on my own or just use them, and load them at the code directly.

I am willing to get all the responsibility for this changes (otherwise I would not have taken this task on myself :) )

**Comment by ik5** [comment]
@sharvit can you help, and review this PR to see that for "bootstrapping" at least the css for now, it is good?

Thank you

**Comment by sharvit** [comment]
@ik5 It looks fine but I wouldn't merge it before I see it is actually doing the job.

Need to load the relevant compiled js and css into your views/layouts and see if it is actually work.
see: 
https://github.com/theforeman/foreman/blob/develop/app/views/layouts/base.html.erb#L13
https://github.com/theforeman/foreman/blob/develop/app/views/layouts/base.html.erb#L18-L19

**Comment by ik5** [comment]
I'ved tested it with SCSS, and it was compiled into the bundle, at this
time, after discussions, the javascript will not be changed, unless needed
to.

**Comment by ohadlevy** [comment]
@ik5 but you still didnt include it in your layout? nor has any content in it, should you move one of the inline css (or all of them) as part of this pr?

**Comment by ik5** [comment]
I want to add them (all of the inline css) in the next stage, this stage
was intended for the setup of the infrastructure.

**Comment by sharvit** [comment]
@ik5 I personally afraid to merge infrastructure and infrastructure-consumer separately. In my experience, many times, there are changes needed to be at the infrastructure.

I think a part of me can never trust a code that never ran.

**Comment by ik5** [comment]
@sharvit, @ohadlevy I'll moved all of the inline css to this PR, and it will not be the first stage, but rather the whole process.

When there will be a need, I'll add also the Javascript support and even React.

**Comment by ohadlevy** [comment]
@ik5 I can't see any of your changes yet? please just let us know when a re-review is needed.

**Comment by ik5** [comment]
@sharvit, @ohadlevy here are my last changes.

File: webpack/stylesheets/main.scss
Comment (lzap) [inline webpack/stylesheets/main.scss:15]: While I understand the whole direction where Foreman core is going, I am not sold on introducing the whole JS/CSS stack into this plugin because of four extracted inline styles. It does not make much sense, discovery has always been quite UX small, four screens actually in total, we are unlikely adding dozen of new screens in this plugin. It was doing fine with the four styles and two JS functions inline.

I am not familiar with these technologies myself, but I do most of reviews, all the release engineering, packaging and troubleshooting upstream and downstream and these four lines can make my life more tough. Therefore I'd like to get volunteers from @theforeman/discovery maintainers for active involvement in packaging and related issues. @ares

File: .eslintrc
Comment (ohadlevy) [inline .eslintrc:1]: technically - for css you dont need eslitnrc, but it maybe ok to leave here.

File: app/views/discovered_mailer/_discovered_host.html.erb
Comment (ohadlevy) [inline app/views/discovered_mailer/_discovered_host.html.erb:4]: can't we find a better css matcher? you can probably take any td inside a discovery table or something like that?

File: app/views/discovered_mailer/discovered_summary.html.erb
Comment (ohadlevy) [inline app/views/discovered_mailer/discovered_summary.html.erb]: core uses a different approach using the roadie-rails gem. if you don't want to do it as part of this pr, please revert these changes?

File: webpack/stylesheets/main.scss
Comment (ohadlevy) [reply webpack/stylesheets/main.scss:15]: @lzap I understand your concern, these changes won't look any different if we would move from inline to the traditional rails asset pipeline.

TBH: using inline style is a hack, and we should try to allow great UX experience, maintainability of the code is one of the first steps towards that. I don't think that packaging webpack assets is any different at this stage from any other plugin...

**Review by ohadlevy — CHANGES_REQUESTED** [review]

File: webpack/stylesheets/main.scss
Comment (ik5) [reply webpack/stylesheets/main.scss:15]: @lzap the "inline" was a pain even to you.

For example the code had stuff like `<% th_style = "color: black ..." %> and reused it in many locations.

On email based HTML it can pass, because it's problematic to write good HTML styled for emails, but using an external styling actually does this in a more web concept rather then using tempting to do the same.

I don't think that helpers or templates should have to "know" what does a color, icon etc should be, but rather what class to use.
The reason is, that it's easier to change style without touching the Ruby/Template constantly, unless there is a change to the class name.


For SCSS, it's a macro language that transfer into CSS rules. It contains variables (that just started to appear in CSS itself), it contains mixin for not repeating the same rule, it contains the ability to create nested rules, and create a "story" in that language to explain the styling for all levels, and more.
And then the language is transpiled into CSS rules, and the macros translated to actual rules, the variables are like place holders etc...

File: app/views/discovered_mailer/_discovered_host.html.erb
Comment (ik5) [reply app/views/discovered_mailer/_discovered_host.html.erb:4]: I was thinking about [bem](http://getbem.com/introduction/), but didn't want to go all the way, because nothing works in that manner here, so even the name convention is not bem, but hints for it.

I do understand why to choose id instead of class for one element, but if I understand correctly the design, each cell might contain different styling, so I didn't want to break that design.

Is my understanding wrong?

File: app/views/discovered_mailer/discovered_summary.html.erb
Comment (ik5) [reply app/views/discovered_mailer/discovered_summary.html.erb]: missed that, sorry, I'm fixing it

Comment (ik5) [reply app/views/discovered_mailer/discovered_summary.html.erb]: Add the change, did not test it yet, I will notify when it is ready

File: app/assets/stylesheets/summary.css
Comment (sharvit) [inline app/assets/stylesheets/summary.css]: What is the reason to put this css here and not in webpack?

**Comment by ik5** [comment]
Testing the current plugin, trying to figure out few things, actually
looking to see how to include it in webpack, but load it only for the email
and not as a link, but rather then a content. So it's for test purpose only.

**Comment by ik5** [comment]
two issues:

1. following discussion with @amirfefer, the `roadie-rails` plugin, does not support the webpack helper.
2. I'm missing here something, but I can't make the css file `email_summary.css` to enter the pipeline.
@sharvit Can you please take a look and maybe see what I'm missing, I worked on it for a day and a half, and I think that it's something small and stupid that I'm missing.
Thanks :)

File: app/views/discovered_mailer/discovered_summary.html.erb
Comment (sharvit) [inline app/views/discovered_mailer/discovered_summary.html.erb]: Can you try it without the `.css`?
Do you actually need turbokinks? isn't it an email template?
```erb
<%= stylesheet_link_tag 'email_summary' %>
```

Comment (ik5) [reply app/views/discovered_mailer/discovered_summary.html.erb]: Removed them both, and still it is not part of the pipeline :(

**Comment by sharvit** [comment]
Try moving it to `stylesheets/foreman_discovery/email_summary.css`

**Comment by ik5** [comment]
now it finds the css, unable to test it yet by hand, working on it

So please DO NOT MERGE YET

**Comment by ik5** [comment]
It works :)

**Comment by lzap** [comment]
You say "stage 1", can you describe all stages, what is the plan? Please include detailed packaging changes required for this to work on RPM/DEB installs.

**Comment by ik5** [comment]
@lzap the discussion here actually changed from "stage1" to the only stage.
I moved all possible CSS to webpack rather then inline CSS, and that's the whole work, that is adding support for webpack, and except for email sending, not using asset pipeline, but rather webpack.

**Comment by lzap** [comment]
Ok, thanks. I need to know what the packaging necessary changes are. If I merge this, will packaging break? If so, let's do package changes first.

**Comment by ik5** [comment]
I'm not sure that I know, that is I know what components I'm using, but not how packaging works for them regarding the way foreman package things.

@ekohl for asset pipeline, and for webpack, what is required for packaging purpose?
The Asset pipeline is used by the action-mailer, and webpack (css at this time) for anything else.

**Comment by tbrisker** [comment]
FWIW, the last time I looked into formatting in HTML mails, some clients didn't support any non-inline styling. This may have changed by now, since it was about 4 years ago, but this should be tested to make sure the formatting works well on at least some common clients (such as Outlook for example)

**Comment by ekohl** [comment]
> @ekohl for asset pipeline, and for webpack, what is required for packaging purpose?

In https://github.com/theforeman/foreman-packaging/commit/2655ffbc47fcd192f61bb900c04279abdd089537 we added it to the template. You can now regenerate a spec file with [add_gem_package.sh](https://github.com/theforeman/foreman-packaging/blob/rpm/develop/add_gem_package.sh) and if the gem contains the webpack directory it should work. The only thing we haven't looked into is additional nodejs dependencies that foreman doesn't have.

**Comment by ik5** [comment]
@tbrisker the email plugin that is used actually takes the asset pipe line and placed it literally as inline css for each class element that exists, so you'll find there also `style="..."`.

@lzap based on @ekohl's  replay, there is no need at this time for anything special and it will take it all. For JS at this time we do not have anything special at all.

**Comment by lzap** [comment]
@ik5 that's not how I interpret @ekohl reply. He says that they made changes to Foreman Plugin SPEC template, foreman_discovery.spec still needs to be changed and tested. And I am not gonna do this, I have no clue how node.js works. I can review and test tho.

**Comment by ik5** [comment]
@lzap the thing is, is that no change to the javascript was made, only moved styling to scss and loading them, and based on @ekohl answer, because it is under `webpack` directory it knows how to use it.

Or am I missing something?

**Comment by ekohl** [comment]
@lzap is correct: we do need to modify the packaging but it's well known which changes need to be made and it should be easy to do.

**Comment by lzap** [comment]
Ok, I am gonna test and merge this after release of discovery for 1.18 foreman, just to be sure we have enough time for fixes. I am not confident in this area. Ping me if I forget after the release.

**Comment by ik5** [comment]
@lzap, please allow me first to finish understanding how to package it (it's not just the release but also a learning curve for me), so for next requirement it will be faster, but not necessarily simpler.

**Comment by lzap** [comment]
No rush, I will be branching off Discovery next week.

**Comment by ekohl** [comment]
Given the huge difficulties webpack introduces in plugins I'd advise any plugin author to hold it off until it stabilizes. This also mixes 2 asset compilation stacks. Is that intentional?

**Comment by ik5** [comment]
@ekohl the two assets are intentional.
The webpack stack is for the web related parts, that is things that are in display on the browser.

The asset pipline, is actually for email, we are using email library that takes the classes names and place at the email body the actual style based on that.
But it does not know how to take it from webpack, only from Rails asset pipeline.

**Comment by lzap** [comment]
Let's revisit this post 1.19 then.

**Comment by lzap** [comment]
Discovery develop is finally green again, please rebase to retest on jenkins.

**Comment by rabajaj0509** [comment]
@ik5 this seems really close, would you like to solve the conflicts and rebase?

**Pull request closed without merging**
